### ISSUE TITLE
❓ [QUESTION] Merge stress support into master

### ISSUE THREAD

**Issue by svandenhaute** [mir-group/nequip#114]
I've been using the stress branch for quite some time and I haven't experienced any issues with it so far. I was wondering whether it is possible to merge this functionality in the master branch?

**Comment by Linux-cpp-lisp** [comment]
Hi @svandenhaute ,

Thanks for the reminder about this, glad to hear that it's been working for you. If I remember correctly, you had tested the `stress` branch and confirmed that the implementation of stress was giving sane answers?

Things are quite busy right now but I will try to get `stress` merged into the latest unstable `develop` as soon as I can; it shouldn't be too complicated. That way you (and others) will be able to use stress while using all the latest features.

**Comment by Linux-cpp-lisp** [comment]
Hi @svandenhaute:

I've implemented stress in a much more complete way in this PR: https://github.com/mir-group/nequip/pull/119 (which is a branch off of our latest `develop`).
We'll be merging this as soon as we finish some internal testing to confirm that everything works as intended, but wanted to let you know as our "beta user" for stress.

Please let me know if you have any issues or thoughts on this implementation. (We may be slow to respond in the next couple of weeks over the winter holidays.)

Thanks!

**Comment by svandenhaute** [comment]
Cool, thanks! I'm also rather busy for the next few weeks but I'll give it a try as soon as possible and report back.

**Comment by Linux-cpp-lisp** [comment]
Great, thanks @svandenhaute --- no rush, and I'll let you know here when we merge it so you can test from the correct branch.

**Comment by svandenhaute** [comment]
@Linux-cpp-lisp what's the status on this? Is the stress-develop branch OK for me to start testing with? Will it be included in the next release?

**Comment by simonbatzner** [comment]
Hi @svandenhaute, I have been stress-testing (badum-tss) the stress-develop branch and everything is working very well, I can get low stress errors! Go ahead and start using it. Yes, this will be included in the next release 

(cc'ing @Linux-cpp-lisp to confirm)

**Comment by Linux-cpp-lisp** [comment]
Yep @svandenhaute! I hadn't merged yet because I had been waiting for @simonbatzner to finish his tests, but they are all looking good so far as he says so I will merge it to `develop` when the rest are done.

I've been regularly merging `develop` into `stress-develop`, so it should be safe to use `stress-develop` as-is, and yes, I plan to merge this before the next release. Thanks!

**Comment by Linux-cpp-lisp** [comment]
@svandenhaute @PythonFZ 

Just a heads up that I've gone ahead and merged `stress-develop` into `develop` to end the merge hell that's starting to accumulate on that branch. Hopefully this makes it easier to use and test.

That said, **stress support is still very much in beta** and, as the warning I added says, please sanity check your results and let us know if you find things that seem strange or wrong.

**Comment by svandenhaute** [comment]
Great news, thanks!

I haven't experienced any issues with it so far. On the contrary, I've found that for the systems I study, NequIP is actually able to predict the virial stress to relatively high accuracy (with MAEs typically below 50 MPa).

**Comment by Linux-cpp-lisp** [comment]
That's great @svandenhaute — one thing I should mention in case (as it sounds like?) you are using `virial` labels and not `stress` ones is that yesterday before merging I changed the sign convention for the predicted virial to
```
virial = -stress x volume
```
See https://github.com/libAtoms/QUIP/issues/227. If you want the previous and, as far as I can tell, less standard convention of 
```
virial = stress x volume
```
you need to set `virial_alternate_sign_convention: True`.

**Comment by tgmaxson** [comment]
Is there any progress on merging this?  For my use case, I need stresses.  I will for now try to use the stress branch but it would be great to get this merged in if its working.

**Comment by tgmaxson** [comment]
Nevermind it seems it has already been merged, how do I go about actually using evaluating the stresses and/or including them in to the loss function (forgive me if my terminology isn't exactly correct,  I haven't been using neural network based codes for too long).

**Comment by Linux-cpp-lisp** [comment]
Hi @tgmaxson ,

Yes, this has been merged both into `develop` and `main` at this point.

> including them in to the loss function

See `configs/minimal_stress.yaml` for a basic example. Note that you can train on `virial` or `stress` depending on what is in your data, but **make sure your data's sign convention aligns with ours** (https://github.com/mir-group/nequip/blob/main/nequip/nn/_grad_output.py#L323-L326). Also be aware that because the `virial` lacks the `1/V` you may want to set the normalization or loss coefficient differently.

> evaluating the stresses

`nequip-evaluate` can handle `stress` and `virial` correctly just like other fields when using `--output` and `--output-fields` (`stress` is output by default, `virial` you have to ask for using `--output-fields`). The `nequip.ase.NequIPCalculator` ASE calculator supports stress. Work on stress support for our LAMMPS plugins is ongoing.
